src/Raven.Client/Documents/InMemoryDocumentSessionOperations.cs

```diff
@@ -644,43 +641,105 @@ private void PrepareForEntitiesDeletion(SaveChangesData result, IDictionary<stri
 
-                    etag = UseOptimisticConcurrency ? etag : null;
-                    result.Entities.Add(value.Entity);
-
-                    /*result.Commands.Add(new DeleteCommandData
+                    result.Commands.Add(new DynamicJsonValue()
                     {
-                        Etag = etag,
-                        Id = key,
-                    });*/
+                        ["Key"] = entity.Value.Id,
+                        ["Method"] = "PUT",
+                        ["Document"] = document
+                    });
+                }
+            }
+            DocumentsByEntity.Clear();
+        }
+
+        protected bool EntityChanged(object entity, DocumentInfo documentMetadata,  out BlittableJsonReaderObject newObj)
+        {
+            //TODO - ConvertEntityToBlittable not good
+            newObj = EntityToBlittable.ConvertEntityToBlittable(documentMetadata.Id, entity, documentMetadata.Metadata);
+            if (documentMetadata == null)
+                return true;
+
+            if (documentMetadata.IgnoreChanges)
+                return false;
+
+            string id;
+            if (GenerateEntityIdOnTheClient.TryGetIdFromInstance(entity, out id) &&
+                string.Equals(documentMetadata.Id, id, StringComparison.OrdinalIgnoreCase) == false)
+                return true;
+
+            // prevent saves of a modified read only entity
+            bool readOnly;
+            documentMetadata.Metadata.TryGet(Constants.Headers.RavenReadOnly, out readOnly);
+            if (readOnly)
+                return false;
+
+            if (documentMetadata.IsNewDocument || documentMetadata.Document == null)
+                return true;
+
+            return compareBlittable(documentMetadata.Document, newObj);
+
+        }
+
+        private static bool compareBlittable(BlittableJsonReaderObject originalBlittable, BlittableJsonReaderObject newBlittable)
+        {
+            var propertiesIds = newBlittable.GetPropertiesByInsertionOrder();
+            if (propertiesIds.Length != originalBlittable.GetPropertiesByInsertionOrder().Length)
+                return true;
+
+            foreach (var propID in propertiesIds)
+            {
+                var propInfo = newBlittable.GetPropertyByIndex(propID);
+
+                object oldPropValue;
+                if (propInfo.Item1 == Constants.Metadata.Key)
+                    continue;
+                originalBlittable.TryGetMember(propInfo.Item1, out oldPropValue);
+
+                if (propInfo.Item3 == BlittableJsonToken.StartObject)
+                {
+                     if (compareBlittable(oldPropValue as BlittableJsonReaderObject, propInfo.Item2 as BlittableJsonReaderObject))
+                        continue;
+                     else
+                        return true;
                 }
 
+                if ((propInfo.Item2 == null ? null : propInfo.Item2.ToString()) != (oldPropValue == null ? null : oldPropValue.ToString()))
```

Comment:
Very bad, you are forcing the value to be a string, which causes allocations, and the value can be big array.
Need to compare on all the values.